Baby Names portfolio project  
Robert Kuyer

Connect to PostgreSQL

In [1]:
import os

%load_ext sql
database_url = os.getenv('database_url_portfolio')
%sql $database_url

Check the time frame of the dataset

In [2]:
%%sql 
SELECT MIN(year), MAX(year)
FROM baby_names

 * postgresql://postgres:***@localhost:5432/PortfolioProject
1 rows affected.


[(1880, 2014)]

Check data entries. Whereas some names are unisex, most names should not show more than 135 year.

In [3]:
%%sql 
SELECT first_name, 
    count(first_name), 
    SUM(num) AS total_amount
FROM baby_names
GROUP BY first_name
HAVING count(first_name) > 200
ORDER BY total_amount DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/PortfolioProject
10 rows affected.


[('James', 270, Decimal('5129096')),
 ('John', 270, Decimal('5106590')),
 ('Robert', 270, Decimal('4816785')),
 ('Michael', 243, Decimal('4330805')),
 ('Mary', 265, Decimal('4130441')),
 ('William', 270, Decimal('4071368')),
 ('David', 253, Decimal('3590557')),
 ('Joseph', 270, Decimal('2580687')),
 ('Richard', 255, Decimal('2564867')),
 ('Charles', 269, Decimal('2376700'))]

Note: this table shows data entry errors in sex (270 occurances for John & Robert), therefore group by sex in coming tables


Select first names that occur every year

In [4]:
%%sql 
SELECT first_name, 
    SUM(num) AS total_amount
FROM baby_names
GROUP BY first_name, sex
HAVING COUNT(first_name) = 135
ORDER BY total_amount DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/PortfolioProject
10 rows affected.


[('James', Decimal('5105919')),
 ('John', Decimal('5084943')),
 ('Robert', Decimal('4796695')),
 ('Michael', Decimal('4309198')),
 ('Mary', Decimal('4115282')),
 ('William', Decimal('4055473')),
 ('David', Decimal('3577704')),
 ('Joseph', Decimal('2570095')),
 ('Richard', Decimal('2555330')),
 ('Charles', Decimal('2364332'))]

Classify names on popularity accross years. Categorizing them as 'Classic', 'Semi-classic', 'Semi-trendy', 'Trendy', or other.  
Thereafter, review trendy names 

In [5]:
%%sql 
WITH classified_names AS (
    SELECT first_name, 
        SUM(num) AS total_amount, 
        COUNT(first_name) AS number_of_years, 
        CASE WHEN COUNT(first_name) > 0 
            AND COUNT(first_name) < 25
            AND sum(num) > COUNT(num) * 100 THEN 'Trendy'
        WHEN COUNT(first_name) >= 25 
            AND COUNT(first_name) < 50 
            AND sum(num) > COUNT(num) * 100 THEN 'Semi-trendy'
        WHEN COUNT(first_name) >= 50 
            AND COUNT(first_NAME) < 100 
            AND sum(num) > COUNT(num) * 75  THEN 'Semi-classic'
        WHEN COUNT(first_name) >= 100 
            AND sum(num) > COUNT(num) * 75  THEN 'Classic' 
        WHEN sum(num) < 200 THEN 'Rare' 
        ELSE 'Other' END AS popularity_type
    FROM baby_names
    GROUP BY first_name, sex
    ORDER BY popularity_type DESC
)
SELECT * 
FROM classified_names
WHERE popularity_type = 'Trendy'
ORDER BY total_amount DESC
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/PortfolioProject
20 rows affected.


[('Nevaeh', Decimal('64713'), 18, 'Trendy'),
 ('Ayden', Decimal('55808'), 24, 'Trendy'),
 ('Jaylen', Decimal('29615'), 24, 'Trendy'),
 ('Jaxson', Decimal('26109'), 24, 'Trendy'),
 ('Kaiden', Decimal('25638'), 23, 'Trendy'),
 ('Jaiden', Decimal('23710'), 24, 'Trendy'),
 ('Maddox', Decimal('21187'), 20, 'Trendy'),
 ('Aniyah', Decimal('20313'), 24, 'Trendy'),
 ('Maximus', Decimal('17659'), 18, 'Trendy'),
 ('Kobe', Decimal('14305'), 21, 'Trendy'),
 ('Caiden', Decimal('14067'), 23, 'Trendy'),
 ('Emerson', Decimal('13225'), 22, 'Trendy'),
 ('Kaydence', Decimal('12345'), 13, 'Trendy'),
 ('Zayden', Decimal('10319'), 15, 'Trendy'),
 ('Yahir', Decimal('9152'), 20, 'Trendy'),
 ('Janiyah', Decimal('9039'), 20, 'Trendy'),
 ('Jax', Decimal('8909'), 20, 'Trendy'),
 ('Kennedi', Decimal('8807'), 24, 'Trendy'),
 ('Mekhi', Decimal('8802'), 20, 'Trendy'),
 ('Kadence', Decimal('8661'), 21, 'Trendy')]

Select the top 10 female names from in the 21st century starting with a z or a q.


In [6]:
%%sql 
SELECT RANK() OVER(ORDER BY SUM(num) DESC) AS rank, 
    first_name, 
    SUM(num) AS total_amount
FROM baby_names
WHERE sex = 'F'
    AND year > 2000
    AND first_name LIKE 'Z%' 
    OR first_name LIKE 'Q%'
GROUP BY first_name
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/PortfolioProject
10 rows affected.


[(1, 'Zoe', Decimal('75701')),
 (2, 'Zoey', Decimal('54287')),
 (3, 'Quinn', Decimal('46282')),
 (4, 'Quentin', Decimal('38471')),
 (5, 'Quinton', Decimal('27448')),
 (6, 'Quincy', Decimal('26060')),
 (7, 'Queen', Decimal('14028')),
 (8, 'Quintin', Decimal('13881')),
 (9, 'Quinten', Decimal('6822')),
 (10, 'Zoie', Decimal('6731'))]

Analyze the expansion of Zoe's: Part 1, which year were most Zoe's born.


In [7]:
%%sql 
SELECT year, 
    first_name, 
    SUM(num) AS amount 
FROM baby_names
WHERE first_name = 'Zoe'
GROUP BY year, first_name
ORDER BY amount DESC
LIMIT 1;

 * postgresql://postgres:***@localhost:5432/PortfolioProject
1 rows affected.


[(2012, 'Zoe', Decimal('6452'))]

Analyze the expansion of Zoe's: Part 2, cummulative Zoe's.


In [8]:
%%sql 
SELECT year, 
    first_name, 
    num AS amount, 
    SUM(num) OVER(ORDER BY year) AS cumulative_zoes
FROM baby_names
WHERE first_name = 'Zoe'
    AND sex = 'F'
ORDER BY year;

 * postgresql://postgres:***@localhost:5432/PortfolioProject
135 rows affected.


[(1880, 'Zoe', 23, Decimal('23')),
 (1881, 'Zoe', 22, Decimal('45')),
 (1882, 'Zoe', 25, Decimal('70')),
 (1883, 'Zoe', 23, Decimal('93')),
 (1884, 'Zoe', 31, Decimal('124')),
 (1885, 'Zoe', 27, Decimal('151')),
 (1886, 'Zoe', 25, Decimal('176')),
 (1887, 'Zoe', 34, Decimal('210')),
 (1888, 'Zoe', 42, Decimal('252')),
 (1889, 'Zoe', 29, Decimal('281')),
 (1890, 'Zoe', 42, Decimal('323')),
 (1891, 'Zoe', 34, Decimal('357')),
 (1892, 'Zoe', 34, Decimal('391')),
 (1893, 'Zoe', 23, Decimal('414')),
 (1894, 'Zoe', 28, Decimal('442')),
 (1895, 'Zoe', 34, Decimal('476')),
 (1896, 'Zoe', 36, Decimal('512')),
 (1897, 'Zoe', 35, Decimal('547')),
 (1898, 'Zoe', 30, Decimal('577')),
 (1899, 'Zoe', 27, Decimal('604')),
 (1900, 'Zoe', 26, Decimal('630')),
 (1901, 'Zoe', 26, Decimal('656')),
 (1902, 'Zoe', 34, Decimal('690')),
 (1903, 'Zoe', 19, Decimal('709')),
 (1904, 'Zoe', 27, Decimal('736')),
 (1905, 'Zoe', 24, Decimal('760')),
 (1906, 'Zoe', 19, Decimal('779')),
 (1907, 'Zoe', 19, Decimal('798')),
 (1908, 'Zoe', 23, Decimal('821')),
 (1909, 'Zoe', 22, Decimal('843')),
 (1910, 'Zoe', 34, Decimal('877')),
 (1911, 'Zoe', 30, Decimal('907')),
 (1912, 'Zoe', 37, Decimal('944')),
 (1913, 'Zoe', 28, Decimal('972')),
 (1914, 'Zoe', 37, Decimal('1009')),
 (1915, 'Zoe', 57, Decimal('1066')),
 (1916, 'Zoe', 65, Decimal('1131')),
 (1917, 'Zoe', 57, Decimal('1188')),
 (1918, 'Zoe', 68, Decimal('1256')),
 (1919, 'Zoe', 53, Decimal('1309')),
 (1920, 'Zoe', 80, Decimal('1389')),
 (1921, 'Zoe', 90, Decimal('1479')),
 (1922, 'Zoe', 64, Decimal('1543')),
 (1923, 'Zoe', 61, Decimal('1604')),
 (1924, 'Zoe', 74, Decimal('1678')),
 (1925, 'Zoe', 73, Decimal('1751')),
 (1926, 'Zoe', 58, Decimal('1809')),
 (1927, 'Zoe', 57, Decimal('1866')),
 (1928, 'Zoe', 64, Decimal('1930')),
 (1929, 'Zoe', 73, Decimal('2003')),
 (1930, 'Zoe', 54, Decimal('2057')),
 (1931, 'Zoe', 77, Decimal('2134')),
 (1932, 'Zoe', 86, Decimal('2220')),
 (1933, 'Zoe', 79, Decimal('2299')),
 (1934, 'Zoe', 79, Decimal('2378')),
 (1935, 'Zoe', 90, Decimal('2468')),
 (1936, 'Zoe', 78, Decimal('2546')),
 (1937, 'Zoe', 72, Decimal('2618')),
 (1938, 'Zoe', 65, Decimal('2683')),
 (1939, 'Zoe', 55, Decimal('2738')),
 (1940, 'Zoe', 54, Decimal('2792')),
 (1941, 'Zoe', 60, Decimal('2852')),
 (1942, 'Zoe', 53, Decimal('2905')),
 (1943, 'Zoe', 40, Decimal('2945')),
 (1944, 'Zoe', 48, Decimal('2993')),
 (1945, 'Zoe', 45, Decimal('3038')),
 (1946, 'Zoe', 66, Decimal('3104')),
 (1947, 'Zoe', 54, Decimal('3158')),
 (1948, 'Zoe', 68, Decimal('3226')),
 (1949, 'Zoe', 81, Decimal('3307')),
 (1950, 'Zoe', 67, Decimal('3374')),
 (1951, 'Zoe', 96, Decimal('3470')),
 (1952, 'Zoe', 100, Decimal('3570')),
 (1953, 'Zoe', 96, Decimal('3666')),
 (1954, 'Zoe', 118, Decimal('3784')),
 (1955, 'Zoe', 120, Decimal('3904')),
 (1956, 'Zoe', 92, Decimal('3996')),
 (1957, 'Zoe', 143, Decimal('4139')),
 (1958, 'Zoe', 151, Decimal('4290')),
 (1959, 'Zoe', 126, Decimal('4416')),
 (1960, 'Zoe', 119, Decimal('4535')),
 (1961, 'Zoe', 123, Decimal('4658')),
 (1962, 'Zoe', 97, Decimal('4755')),
 (1963, 'Zoe', 81, Decimal('4836')),
 (1964, 'Zoe', 79, Decimal('4915')),
 (1965, 'Zoe', 78, Decimal('4993')),
 (1966, 'Zoe', 118, Decimal('5111')),
 (1967, 'Zoe', 82, Decimal('5193')),
 (1968, 'Zoe', 86, Decimal('5279')),
 (1969, 'Zoe', 90, Decimal('5369')),
 (1970, 'Zoe', 134, Decimal('5503')),
 (1971, 'Zoe', 121, Decimal('5624')),
 (1972, 'Zoe', 96, Decimal('5720')),
 (1973, 'Zoe', 121, Decimal('5841')),
 (1974, 'Zoe', 106, Decimal('5947')),
 (1975, 'Zoe', 126, Decimal('6073')),
 (1976, 'Zoe', 97, Decimal('6170')),
 (1977, 'Zoe', 121, Decimal('6291')),
 (1978, 'Zoe', 94, Decimal('6385')),
 (1979, 'Zoe', 108, Decimal('6493')),
 (1980, 'Zoe', 102, Decimal('6595')),
 (1981, 'Zoe', 110, Decimal('6705')),
 (1982, 'Zoe', 130, Decimal('6835')),
 (1983, 'Zoe', 174, Decimal('7009')),
 (1984, 'Zoe', 201, Decimal('7210')),
 (1985, 'Zoe', 194, Decimal('7404')),
 (1986, 'Zoe', 213, Decimal('7617')),
 (1987, 'Zoe', 247, Decimal('7864')),
 (1988, 'Zoe', 240

Select the most popular male and female name for each year.


In [9]:
%%sql 
WITH ranked_names AS (
    SELECT year,
        first_name,
        sex,
        num,
        ROW_NUMBER() OVER (PARTITION BY year, sex ORDER BY num DESC) AS rn
    FROM baby_names
)
SELECT year,
    MAX(CASE WHEN sex = 'M' AND rn = 1 THEN first_name END) AS male_name,
    MAX(CASE WHEN sex = 'M' AND rn = 1 THEN num END) AS max_male_num,
    MAX(CASE WHEN sex = 'F' AND rn = 1 THEN first_name END) AS female_name,
    MAX(CASE WHEN sex = 'F' AND rn = 1 THEN num END) AS max_female_num
FROM ranked_names
GROUP BY year
ORDER BY year DESC;

 * postgresql://postgres:***@localhost:5432/PortfolioProject
135 rows affected.


[(2014, 'Noah', 19144, 'Emma', 20799),
 (2013, 'Noah', 18179, 'Sophia', 21147),
 (2012, 'Jacob', 19030, 'Sophia', 22267),
 (2011, 'Jacob', 20331, 'Sophia', 21816),
 (2010, 'Jacob', 22082, 'Isabella', 22883),
 (2009, 'Jacob', 21144, 'Isabella', 22278),
 (2008, 'Jacob', 22568, 'Emma', 18799),
 (2007, 'Jacob', 24252, 'Emily', 19345),
 (2006, 'Jacob', 24830, 'Emily', 21389),
 (2005, 'Jacob', 25814, 'Emily', 23928),
 (2004, 'Jacob', 27871, 'Emily', 25025),
 (2003, 'Jacob', 29615, 'Emily', 25685),
 (2002, 'Jacob', 30556, 'Emily', 24457),
 (2001, 'Jacob', 32527, 'Emily', 25052),
 (2000, 'Jacob', 34465, 'Emily', 25952),
 (1999, 'Jacob', 35344, 'Emily', 26537),
 (1998, 'Michael', 36613, 'Emily', 26179),
 (1997, 'Michael', 37543, 'Emily', 25731),
 (1996, 'Michael', 38364, 'Emily', 25149),
 (1995, 'Michael', 41403, 'Jessica', 27938),
 (1994, 'Michael', 44465, 'Jessica', 32116),
 (1993, 'Michael', 49553, 'Jessica', 34984),
 (1992, 'Michael', 54377, 'Ashley', 38453),
 (1991, 'Michael', 60781, 'Ashley', 43482),
 (1990, 'Michael', 65274, 'Jessica', 46466),
 (1989, 'Michael', 65374, 'Jessica', 47882),
 (1988, 'Michael', 64116, 'Jessica', 51530),
 (1987, 'Michael', 63631, 'Jessica', 55985),
 (1986, 'Michael', 64184, 'Jessica', 52667),
 (1985, 'Michael', 64887, 'Jessica', 48346),
 (1984, 'Michael', 67722, 'Jennifer', 50559),
 (1983, 'Michael', 67986, 'Jennifer', 54336),
 (1982, 'Michael', 68204, 'Jennifer', 57109),
 (1981, 'Michael', 68752, 'Jennifer', 57041),
 (1980, 'Michael', 68666, 'Jennifer', 58385),
 (1979, 'Michael', 67728, 'Jennifer', 56715),
 (1978, 'Michael', 67153, 'Jennifer', 56311),
 (1977, 'Michael', 67611, 'Jennifer', 58960),
 (1976, 'Michael', 66971, 'Jennifer', 59479),
 (1975, 'Michael', 68457, 'Jennifer', 58185),
 (1974, 'Michael', 67577, 'Jennifer', 63111),
 (1973, 'Michael', 67864, 'Jennifer', 62454),
 (1972, 'Michael', 71426, 'Jennifer', 63606),
 (1971, 'Michael', 77611, 'Jennifer', 56786),
 (1970, 'Michael', 85321, 'Jennifer', 46160),
 (1969, 'Michael', 85227, 'Lisa', 45032),
 (1968, 'Michael', 82019, 'Lisa', 49532),
 (1967, 'Michael', 82454, 'Lisa', 52430),
 (1966, 'Michael', 80002, 'Lisa', 56913),
 (1965, 'Michael', 81048, 'Lisa', 60268),
 (1964, 'Michael', 82674, 'Lisa', 54275),
 (1963, 'Michael', 83795, 'Lisa', 56039),
 (1962, 'Michael', 85046, 'Lisa', 46083),
 (1961, 'Michael', 86916, 'Mary', 47667),
 (1960, 'David', 85931, 'Mary', 51479),
 (1959, 'Michael', 85272, 'Mary', 54473),
 (1958, 'Michael', 90519, 'Mary', 55850),
 (1957, 'Michael', 92709, 'Mary', 61096),
 (1956, 'Michael', 90633, 'Mary', 61751),
 (1955, 'Michael', 88283, 'Mary', 63160),
 (1954, 'Michael', 88485, 'Mary', 68001),
 (1953, 'Robert', 86135, 'Mary', 64337),
 (1952, 'James', 87083, 'Linda', 67096),
 (1951, 'James', 87175, 'Linda', 73933),
 (1950, 'James', 86221, 'Linda', 80431),
 (1949, 'James', 86856, 'Linda', 91010),
 (1948, 'James', 88596, 'Linda', 96205),
 (1947, 'James', 94755, 'Linda', 99680),
 (1946, 'James', 87425, 'Mary', 67463),
 (1945, 'James', 74450, 'Mary', 59284),
 (1944, 'James', 76944, 'Mary', 62463),
 (1943, 'James', 80248, 'Mary', 66171),
 (1942, 'James', 77173, 'Mary', 63238),
 (1941, 'James', 66719, 'Mary', 58031),
 (1940, 'James', 62472, 'Mary', 56195),
 (1939, 'Robert', 59640, 'Mary', 54900),
 (1938, 'Robert', 62262, 'Mary', 56211),
 (1937, 'Robert', 61821, 'Mary', 55640),
 (1936, 'Robert', 58486, 'Mary', 54363),
 (1935, 'Robert', 56514, 'Mary', 55067),
 (1934, 'Robert', 55826, 'Mary', 56918),
 (1933, 'Robert', 54205, 'Mary', 55493),
 (1932, 'Robert', 59248, 'Mary', 59871),
 (1931, 'Robert', 60497, 'Mary', 60293),
 (1930, 'Robert', 62143, 'Mary', 64134),
 (1929, 'Robert', 59802, 'Mary', 63508),
 (1928, 'Robert', 60680, 'Mary', 66866),
 (1927, 'Robert', 61664, 'Mary', 70632),
 (1926, 'Robert', 61118, 'Mary', 67823),
 (1925, 'Robert', 60903, 'Mary', 70601),
 (1924, 'Robert', 60792, 'Mary', 73527),
 (1923, 'John', 57471, 'Mary', 71634),
 (1922, 'John', 57278, 'Mary', 72171),
 (1921, 'John', 58228, 'Mary', 73985),
 (1920, 'John', 5

Select the name that was most popular most often

In [10]:
%%sql 
WITH T AS (
    SELECT b.year, 
        b.first_name, 
        b.num
    FROM baby_names AS b
    INNER JOIN (
        SELECT year, 
        MAX(num) AS max_num
    FROM baby_names
    GROUP BY year, sex) AS sub
    ON sub.year = b.year AND sub.max_num = num
    ORDER BY year DESC
)
SELECT first_name,
    COUNT(first_name) AS count_top_name
FROM T
GROUP BY first_name
ORDER BY count_top_name DESC;

 * postgresql://postgres:***@localhost:5432/PortfolioProject
17 rows affected.


[('Mary', 76),
 ('John', 44),
 ('Michael', 44),
 ('Robert', 17),
 ('Jennifer', 15),
 ('Jacob', 14),
 ('James', 13),
 ('Emily', 12),
 ('Jessica', 9),
 ('Lisa', 8),
 ('Linda', 6),
 ('Sophia', 3),
 ('Isabella', 2),
 ('Emma', 2),
 ('Noah', 2),
 ('Ashley', 2),
 ('David', 1)]